pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SplitSegmentCommitter.java

```diff
@@ -0,0 +1,87 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.manager.realtime;
+
+import java.io.File;
+import org.apache.pinot.common.protocols.SegmentCompletionProtocol;
+import org.apache.pinot.core.segment.index.loader.IndexLoadingConfig;
+import org.apache.pinot.server.realtime.ServerSegmentCompletionProtocolHandler;
+import org.slf4j.Logger;
+
+
+/**
+ * Sends segmentStart, segmentUpload, & segmentCommitEnd to the controller
+ * If that succeeds, swap in-memory segment with the one built.
+ */
+public class SplitSegmentCommitter implements SegmentCommitter {
+  private SegmentCompletionProtocol.Request.Params _params;
+  private ServerSegmentCompletionProtocolHandler _protocolHandler;
+  private SegmentCompletionProtocol.Response _prevResponse;
+  private IndexLoadingConfig _indexLoadingConfig;
+
+  private Logger _segmentLogger;
+
+  public SplitSegmentCommitter(Logger segmentLogger, ServerSegmentCompletionProtocolHandler protocolHandler,
+      IndexLoadingConfig indexLoadingConfig, SegmentCompletionProtocol.Request.Params params, SegmentCompletionProtocol.Response prevResponse) {
+    _segmentLogger = segmentLogger;
+    _protocolHandler = protocolHandler;
+    _indexLoadingConfig = indexLoadingConfig;
+    _params = params;
+    _prevResponse = prevResponse;
+  }
+
+  @Override
+  public SegmentCompletionProtocol.Response commit(long currentOffset, int numRowsConsumed, LLRealtimeSegmentDataManager.SegmentBuildDescriptor segmentBuildDescriptor) {
+    final File segmentTarFile = new File(segmentBuildDescriptor.getSegmentTarFilePath());
+
+    SegmentCompletionProtocol.Response segmentCommitStartResponse = _protocolHandler.segmentCommitStart(_params);
+    if (!segmentCommitStartResponse.getStatus()
+        .equals(SegmentCompletionProtocol.ControllerResponseStatus.COMMIT_CONTINUE)) {
+      _segmentLogger.warn("CommitStart failed  with response {}", segmentCommitStartResponse.toJsonString());
+      return SegmentCompletionProtocol.RESP_FAILED;
+    }
+
+    SegmentCompletionProtocol.Request.Params params = new SegmentCompletionProtocol.Request.Params();
+    params.withInstanceId(_params.getInstanceId()).withOffset(currentOffset).withSegmentName(_params.getSegmentName());
+    SegmentCompletionProtocol.Response segmentCommitUploadResponse =
+        _protocolHandler.segmentCommitUpload(params, segmentTarFile, _prevResponse.getControllerVipUrl());
+    if (!segmentCommitUploadResponse.getStatus()
+        .equals(SegmentCompletionProtocol.ControllerResponseStatus.UPLOAD_SUCCESS)) {
+      _segmentLogger.warn("Segment upload failed with response {}", segmentCommitUploadResponse.toJsonString());
+      return SegmentCompletionProtocol.RESP_FAILED;
+    }
+
+    _params.withOffset(currentOffset).withSegmentLocation(segmentCommitUploadResponse.getSegmentLocation())
```

Comment:
curent offset and numrows consumed should already be set, and is not going to change between first and this step. We only need to add segmentLocation

Reply:
Because I wanted to maintain the integrity of the previous code, I created a new Params object to wipe state - therefore, for the params object passed in I will keep this for now.